### ISSUE TITLE
Decorating a static listener with @on_trait_change causes problems

### ISSUE THREAD

**Issue by jwiggins** [enthought/traits#527]
This minimal example fails on Python 3.6.0 and Traits 5.1.2:
```python
from traits.api import HasTraits, List, on_trait_change

class TestBase(HasTraits):
    foo = List()

    @on_trait_change('foo_items')
    def _foo_changed(self, obj, trait_name, undefined, event):
        pass

class PatchedBaseClass(TestBase):
    pass
```

Here's the error:
```
Traceback (most recent call last):
  File "test.py", line 13, in <module>
    class PatchedBaseClass(TestBase):
  File "/Users/jwiggins/.edm/envs/edm3/lib/python3.6/site-packages/traits/has_traits.py", line 489, in __new__
    class_name, bases, class_dict, is_category=False
  File "/Users/jwiggins/.edm/envs/edm3/lib/python3.6/site-packages/traits/has_traits.py", line 831, in update_traits_class_dict
    _add_notifiers(trait._notifiers(1), handlers)
  File "/Users/jwiggins/.edm/envs/edm3/lib/python3.6/site-packages/traits/has_traits.py", line 429, in _add_notifiers
    handler = StaticTraitChangeNotifyWrapper(handler)
  File "/Users/jwiggins/.edm/envs/edm3/lib/python3.6/site-packages/traits/trait_notifiers.py", line 373, in __init__
    % (handler.__name__, arg_count)
traits.trait_errors.TraitNotificationError: Invalid number of arguments for the static anytrait change notification handler: _foo_changed. A maximum of 4 arguments is allowed, but 5 were specified.
```

**Comment by mdickinson** [comment]
I'd *expect* combining the `@on_trait_change` decorator with static handler magically-named methods to cause problems, so I don't consider that a bug. The surprise, and the bit that needs investigation, is that it appears to work fine up until the moment where you subclass.

**Comment by mdickinson** [comment]
Question: what's the behaviour we should expect here? Should this be an error? A warning? Should the `on_trait_change` decorator nullify the magic of the `*_changed` naming?

**Comment by rahulporuri** [comment]
> Should this be an error? A warning?

A warning in one release and then an error in a subsequent release?

> Should the on_trait_change decorator nullify the magic of the *_changed naming?

That sounds like the right thing to do - while raising a `RuntimeWarning`.

**Comment by mdickinson** [comment]
> That sounds like the right thing to do - while raising a `RuntimeWarning`.

Agreed.

**Comment by jwiggins** [comment]
Should creation of a subclass still fail?

**Comment by mdickinson** [comment]
> Should creation of a subclass still fail?

No; that part is clearly a bug.

**Comment by kitchoi** [comment]
>> Should the on_trait_change decorator nullify the magic of the *_changed naming?
> That sounds like the right thing to do - while raising a RuntimeWarning.


Do we need the same behaviour for this scenario?
```
@observe("name")
def _name_changed(self, name):
    pass
```

Not sure if this should be a separate issue... At this point it seems to me that the outcome of this issue will affect `observe` if we want consistent behaviour. And it looks like the check will happen in the same place anyway.

**Comment by mdickinson** [comment]
Yes, I'd expect `@observe` to similarly nullify magic naming (again, possibly with a `RuntimeWarning`).

**Comment by mdickinson** [comment]
I can no longer reproduce the issue on the main branch.

From a git bisect, this was apparently fixed in commit f65163e8cfbedfe4f1f1f7bda990a1247ea00798 (PR #638), though it's not clear exactly what part of that PR caused the fix.

**Comment by jwiggins** [comment]
So Traits 6.0.0 and after.

**Comment by mdickinson** [comment]
The relevant change was to `_get_def` in #638: prior to #638, we were using `six.get_unbound_function`, here: https://github.com/enthought/traits/blob/115ea29a372d1a4a3724d21b3198475f1e12f60d/traits/has_traits.py#L211. This meant that we picked up the `_changed` method for subclasses, but not for the main class.

I'm satisfied that the bug no longer exists in main, but this issue is probably still worth a regression test.

**Comment by mdickinson** [comment]
Hmm. There's still a problem with `observe`.

**Comment by mdickinson** [comment]
> There's still a problem with `observe`.

Fix in #1560, along with tests that would have failed for the original issue.
